### ISSUE TITLE
Compute areas

### ISSUE THREAD

**Issue by JimHokanson** [openworm/movement_validation#127]
In `pre_features.py` we need the area to be calculated. Methods used in the eccentricity feature calculation might be useful, although it would be nice if a quick fill algorithm could be found that works.

The approach is to find "pixels" that are within the the given boundaries. The area is then some scalar of the number of pixels (Depends on spatial sampling density).

This result could then be held onto for use by the eccentricity code (since it is nearly the same code twice).

Run the code setup using [`test_normalized_worm_creation.py`](https://github.com/openworm/movement_validation/blob/master/examples/test_normalized_worm_creation.py). Running this function will call `pre_features.py`.  Feel free to temporarily comment out the other aspects of the code to focus on writing the area code.

**Comment by MichaelCurrie** [comment]
I'll elaborate with more details here.  The exact method that needs to be coded is `computeAreas`.  Here's the stub method right now:

```python
    @staticmethod
    def computeAreas(vulva_contour, non_vulva_contour):
        """
        Compute the areas of various parts of the worm, for each frame of 
        video, from a heterocardinal contour.

        Parameters
        -------------------------
        vulva_contour: a list of numpy arrays.
            The list is of the frames.
        non_vulva_contour: same
        
        Returns
        -------------------------
        (head_area, tail_area, vulva_area, non_vulva_area): tuple
            head_area is a numpy array of scalars, giving the area in each 
            frame of video.  so if there are 4000 frames, head_area would have
            shape (4000,)
        
        """
        # TODO: this method is currently unimplemented!
        
        # The old method was:
        # Using known transition regions, count the # of 'on' pixels in
        # the image. Presumably we would use something more akin
        # to the eccentricity feature code
        # - @JimHokanson        

        return (None, None, None, None)




Also, take note of the suggestion in #130 for a technique.

**Comment by ver228** [comment]
Hello, do we really need the area per worm segment? It is only used in worm_features to calculate the worm total area. If one is interested in only the total size we can use this http://mathworld.wolfram.com/PolygonArea.html

I would make the function like this a function:

def computeAreas(vulva_contour, non_vulva_contour):
    contour = np.vstack((vulva_contour, non_vulva_contour[::-1,:])) 
    signed_area = np.sum(contour[:-1,0]*contour[1:,1]-contour[1:,0]*contour[:-1,1])
    return np.abs(signed_area)/2

**Comment by JimHokanson** [comment]
@ver228 

Good point. I was not aware of this technique. The sub areas could also be calculated in this manner by creating sub-polygons. These areas might be used somewhere in the worm parsing verification as well ...

**Comment by ver228** [comment]
@JimHokanson 
I agree, the subareas shouldn't be not too difficult to calculate, but do we really need them? I can only find them being used in two files. 1) NormalizedWorm where it is used to parse the data from Schafer's lab, 2) and worm_features where the subareas are added to obtain the worm area. I personally thing it would be less convoluted to only keep one field for the total worm area. Segworm do use the subareas to filter for bad skeletons, but for my experience from the code, that part is not particularly useful.

**Comment by JimHokanson** [comment]
@ver228 I'm in agreement. My only point was on the preprocessing step, but if that isn't that useful, then I don't think the code should be implemented.

Given the information you have provided I think it is probably best if we remove the individual areas from the old version and only propagate the total area, and then implement the total area in the new code.

**Comment by MichaelCurrie** [comment]
Hi Avelino, you make a good point that in no feature do we actually need area broken down into parts.  The only feature that uses these areas is morphology.area, which as you say just adds up the four areas to get the full worm area (which actually doesn't make sense to me since vulva_area+non_vulva_area should actually be sufficient by my assumed definition of those terms to get the full worm area, but that's an aside..)

So our problem reduces to just needing to calculate the area of the whole worm.  You suggest the [Shoelace Formula](http://en.wikipedia.org/wiki/Shoelace_formula) to accomplish this:

![image](https://cloud.githubusercontent.com/assets/5267078/7994620/064638e0-0acc-11e5-8e2b-fbce6bb947ac.png)

I've made a few changes to your code:

```Python
    @staticmethod
    def computeArea(vulva_contour, non_vulva_contour):
        """
        Compute the area of the worm, for each frame of video, from a 
        normalized contour.

        Parameters
        -------------------------
        vulva_contour: a (49,2,n)-shaped numpy array
        non_vulva_contour: same
        
        Returns
        -------------------------
        area: float
            A numpy array of scalars, giving the area in each 
            frame of video.  so if there are 4000 frames, area would have
            shape (4000,)
        
        """
        # We need the contour points all organized in order around the worm:
        # - vulva_contour, non_vulva_contour have shape (49, 2, n).
        # - contour has shape (97, 2, n).
        contour = np.concatenate((vulva_contour[:, :, :], 
                                  non_vulva_contour[-2:0:-1, :,:]))
    
        # We do this to avoid a RuntimeWarning taking the nanmean of 
        # frames with nothing BUT nan entries: for those frames nanmean 
        # returns nan (correctly) but still raises a RuntimeWarning.
        with warnings.catch_warnings():
            warnings.simplefilter('ignore', category=RuntimeWarning)
            contour_mean = np.nanmean(contour, 0, keepdims=False)
    
        # Centre the contour about the origin for each frame
        # this is technically not necessary but it shrinks the magnitude of 
        # the coordinates we are about to multiply for a potential speedup
        contour -= contour_mean
    
        # We want a new 3D array, where all the points (i.e. axis 0) 
        # are shifted forward by one and wrapped.
        # That is, for a given frame:
        #  x' sub 1 = x sub 2, 
        #  x' sub 2 = x sub 3, 
        #  ...,
        #  x` sub n = x sub 1.     (this is the "wrapped" index)
        # Same for y.
        contour_plus_one = contour.take(range(1,np.shape(contour)[0]+1),
                                        mode='wrap', axis=0)
    
        # Now we use the Shoelace formula to calculate the area of a simple
        # polygon for each frame.
        # Credit to Avelino Javer for suggesting this.
        signed_area = np.nansum(contour[:,0,:]*contour_plus_one[:,1,:] - \
                                contour[:,1,:]*contour_plus_one[:,0,:], 0) / 2
    
        # Frames where the contour[:,:,k] is all NaNs will result in a 
        # signed_area[k] = 0.  We must replace these 0s with NaNs.
        signed_area[np.flatnonzero(np.isnan(contour[0,0,:]))] = np.NaN

        # TODO: convert square pixels to square microns
        pass    
    
        return np.abs(signed_area)
```

**Comment by MichaelCurrie** [comment]
OK so I've refactored so that NormalizedWorm just has a variable, `area`, and when loading from the Schafer file, the factory method itself converts to our new simpler `area` variable:

```Python
    @classmethod
    def from_schafer_file_factory(cls, data_file_path):
        # ...
        nw.area = nw.head_area + nw.tail_area + nw.vulva_area + nw.non_vulva_area
```

And now the features code for morphology is just a simple assignment from the NormalizedWorm value:

```Python
class WormMorphology(object):
    # ...
    def __init__(self, features_ref):
        # ...
        self.area = nw.area
```

I added some temporary code to compare area values into [`test_normalized_worm_creation.py`](https://github.com/openworm/movement_validation/blob/master/examples/test_normalized_worm_creation.py):

```Python
# ...

# Compare area calculations
plt.title('Worm Area')    
area_loaded, = plt.plot(nw.area, 'r', label='Loaded')
area_calculated, = plt.plot(nw_calculated.area, 'b', label='Calculated')
plt.legend(handles=[area_loaded, area_calculated])
plt.show()
```
![image](https://cloud.githubusercontent.com/assets/5267078/7996845/3cce0fee-0ae2-11e5-8965-d08b72bb5af7.png)

It's still off by a scaling factor, which I believe might be the pixels-to-microns conversion that I haven't done.

**Comment by JimHokanson** [comment]
@MichaelCurrie The data should be scaled in microns already. What's the ratio of those 2 lines?

**Comment by JimHokanson** [comment]
https://github.com/openworm/movement_validation/blob/master/jim_matlab_testing/createCleanPreNormData.m

The link above shows the code used to create the data that we are working with. Note that it does the conversion from pixels to microns.

**Comment by MichaelCurrie** [comment]
Hmm.  The calculated area / loaded area is
mean 1.09479022485
min 1.08466231002
max 1.10288471017

![image](https://cloud.githubusercontent.com/assets/5267078/7997255/2bd1bfc4-0ae7-11e5-8125-d80c21637b64.png)

calculated area - loaded area is 

mean 6076.8487679
min 5496.68115274
max 6637.8637457

![image](https://cloud.githubusercontent.com/assets/5267078/7997274/61262ba6-0ae7-11e5-95c7-39cfb9110e48.png)

OK so it doesn't appear that there is a fixed ratio.  It could be that perhaps I'm dropping one polygon point at the end or something, and that's cutting the area calculation by a nearly fixed but still variable ratio at each frame.  I'll investigate that next.

![image](https://cloud.githubusercontent.com/assets/5267078/7997283/8ee53afa-0ae7-11e5-988b-9e2fb04244e1.png)

**Comment by ver228** [comment]
Hello guys,

I made a comparison between using the shoelace formula and filling the worm contour and counting the pixels. The results are practically the same. I think the problem is in the original procedure of dividing the worm into segments.

I am using opencv to fill the worm. I find this library very helpful, and I think it the project will benefit if we start to use it. For example, the eccentricity calculation will be faster. He is how to install it for python 3. http://luigolas.com/blog/2014/09/15/install-opencv3-with-python-3-mac-osx/

```python
# -*- coding: utf-8 -*-
"""
Created on Fri Jun  5 08:11:59 2015

@author: ajaver
"""
import cv2
import numpy as np
import matplotlib.pylab as plt

import sys
import os

sys.path.append('../../movement_validation')
from movement_validation import user_config, NormalizedWorm

# Set up the necessary file paths for file loading
#----------------------
base_path = os.path.abspath(user_config.EXAMPLE_DATA_PATH)

matlab_generated_file_path = os.path.join(
    base_path,'example_video_feature_file.mat')

data_file_path = os.path.join(base_path,"example_video_norm_worm.mat")

# OPENWORM
#----------------------
# Load the normalized worm from file
nw = NormalizedWorm.from_schafer_file_factory(data_file_path)

#Since the first and the last points of vulva and non_vulva contour are equal
#This add two redundant points. The last one is useful to shift the data (close the contour)
#in using the shoelace formula. The one in the middle will not be included in the 
#calculation since it correspond to two equal consecutive points, whos contribution
#will be cancel out. z = x(n-1)*y(n) - x(n)*y(n-1) if x(n) = x(n-1) and y(n) = y(n-1), z = 0
contour = np.concatenate((nw.vulva_contour, nw.non_vulva_contour[::-1,:,:])) 

#this area is signed. It is positive if the contour is going clockwise, and negative if it is counter-clockwise
signed_area = np.sum(contour[:-1,0,:]*contour[1:,1,:]-\
contour[1:,0,:]*contour[:-1,1,:], axis=0)/2

#get the absolute value and obtain the real area
area_shoelace = np.abs(signed_area)


#now let's calculate the area by creating a binary image with a worm image
tot_cnt = nw.vulva_contour.shape[2]
area_filled = np.full(tot_cnt, np.nan);
for ii in range(tot_cnt):
    worm_cnt = contour[:,:,ii];
    
    #stop if the contour is not valid
    if np.any(np.isnan(worm_cnt)):
        continue
    
    #the minimum value of the contour will be the image upper-left corner
    corner = np.min(worm_cnt, axis=0)
    worm_cnt = worm_cnt-corner
    
    #after subracting the corner the maximum contour values will be the other corner
    im_size = np.max(worm_cnt, axis=0)[::-1] #the coordinate systems used by opencv is different from the numpy's one
    
    #TODO: it might be necessary to rescale the contour so the sampling is the same, in all cases
    
    im_dum = np.zeros(im_size)
    
    #draw contours
    cv2.drawContours(im_dum, [worm_cnt.astype(np.int32)], 0, 1, -1)
    area_filled[ii] = np.sum(im_dum)
    
    #example plot
    if ii == 4624:
        plt.figure()
        plt.imshow(im_dum)
        plt.plot(worm_cnt[:,0], worm_cnt[:,1], 'g')
        plt.savefig('Filled_area_example.png

')

plt.figure()
plt.plot(area_shoelace, label ='Shoelace method')
plt.plot(area_filled, label = 'Drawing the polygon')
plt.legend()
plt.ylabel('Area')
plt.ylabel('Frame Number')
plt.savefig('Shoelace_vs_Filled_area.png')

```
![filled_area_example](https://cloud.githubusercontent.com/assets/8364368/8001922/88906e6a-0b62-11e5-806c-6197dfbb9cc3.png)
![shoelace_vs_filled_area](https://cloud.githubusercontent.com/assets/8364368/8001932/93347082-0b62-11e5-8d7b-52b19f6e41c1.png)

**Comment by MichaelCurrie** [comment]
Thanks @ver228, looks great.  I added your code to a [test_areas.py](https://github.com/openworm/movement_validation/blob/master/examples/test_areas.py).

Right now I can't actually run it since I can't get OpenCV working with Python 3 bindings.  The version 3.0 available for download on the OpenCV website for Windows has two `.pyd` files you are supposed to put in your `site-packages` folder, but they are both for 2.7.  Apparently they haven't done us the favour of building a Python 3 `.pyd` right in the install self-extractor!  So instead I've cast about and found this on [StackOverflow](http://stackoverflow.com/questions/20953273/install-opencv-for-python-3-3), and I plan to try it tomorrow.  Fingers crossed.

**Comment by MichaelCurrie** [comment]
Unfortunately I still can't get a compiled `cv2.pyd` file that is for Windows, x64, with Python 3.4 bindings.

I posted about my problems at [StackOverflow](http://stackoverflow.com/questions/29747195/building-opencv-3-0-0-beta-windows-python-3-4).  Suggestions welcome, or if someone could make such a `cv2.pyd` file available for download that would be ideal!

**Comment by JimHokanson** [comment]
Does this not work:
http://www.lfd.uci.edu/~gohlke/pythonlibs/#opencv

OpenCV just updated with Python 3 support two days ago I believe.

**Comment by MichaelCurrie** [comment]
Of course!  I think you might be right.  Let me try that.

p.s. OpenCV had a "Release Candidate" version of 3.0 available since April 24, but yes it appears they just released the production 3.0 version on 4 June 2015.
![image](https://cloud.githubusercontent.com/assets/5267078/8020941/a3e973ae-0c49-11e5-94bf-81a859ad1247.png)

**Comment by MichaelCurrie** [comment]
Worked.  Marvelous.  Thanks Jim
